### PR TITLE
NAS-116269 / 22.12 / validate user/group entries can execute on path in setacl

### PR DESCRIPTION
**Pull request by anodos325**
One of the most common permissions misconfigurations by
end-users is to remove execute on a parent directory path.

This PR validates during setacl that all users and groups in
the proposed ACL have execute access(path, F_OK) on path
leading up to share. Walk up each path component so that
we can identify the exact problem path and give it to the
user as a sort of "hey you need to fix this here path"
message.

### CODE REVIEW

**Comment by anodos325** [comment]
Strategy for validating user is to get the user object. For validating group access, we synthesize one with a non-existent user and the group in question set as primary group.

File: src/middlewared/middlewared/plugins/filesystem_/perm_check.py [hunk 1]
```diff
@@ -25,6 +25,62 @@ def check_access(path: str, check_perms: dict) -> bool:
 
 class FilesystemService(Service):
 
+    @private
+    def generate_user_details(self, id_type, xid):
+        if id_type not in ['USER', 'GROUP']:
```
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/filesystem_/perm_check.py:30]: We can just perhaps use accepts schema ?

File: src/middlewared/middlewared/plugins/filesystem_/perm_check.py [hunk 2]
```diff
@@ -56,4 +112,60 @@ def can_access_as_user(self, username, path, perms):
...
+
+    @private
+    def check_path_execute(self, path, id_type, xid):
+        user_details = self.generate_user_details(id_type, xid)
+        if user_details is None:
+            # User or group does not exist on server.
+            # This can happen for a variety of reasons that are potentially
+            # acceptable (or better than alternative of changing permissions).
+            # Hence, skip validation.
+            self.logger.trace('%s %s does not exist. Skipping validation',
```
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/filesystem_/perm_check.py:125]: `xid` is an integer ? We might want to use `%r` or `%d` ?

File: src/middlewared/middlewared/plugins/filesystem_/perm_check.py [hunk 3]
```diff
@@ -56,4 +112,60 @@ def can_access_as_user(self, username, path, perms):
...
+                id_info['id_type'] = 'USER'
+                id_info['xid'] = uid
+
+            elif entry['tag'] in ('group@', 'GROUP_OBJ'):
+                id_info['id_type'] = 'USER'
+                id_info['xid'] = gid
+
+            self.check_path_execute(path, id_info['id_type'], id_info['xid'])
+
+        return
```
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/filesystem_/perm_check.py:171]: This is not required

File: src/middlewared/middlewared/plugins/filesystem_/acl_linux.py [hunk 4]
```diff
@@ -382,6 +382,35 @@ def getacl(self, path, simplified, resolve_ids):
...
+            for entry in json_verrors:
+                for schema, err in entry.items():
+                    verrors.add(f'filesystem_acl.{schema.replace("acl", "dacl")}', err)
+
+            verrors.check()
+
+        elif setacl.returncode != 0:
+            raise CallError(setacl.stderr.decode())
+
+        return
```
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/filesystem_/acl_linux.py:412]: This is not required

**Review by sonicaj — COMMENTED** [review]
Just some minor nits :) 

Btw can we write some integration tests for this please ?

File: src/middlewared/middlewared/plugins/filesystem_/perm_check.py (hunk 1 above)
Comment (anodos325) [reply src/middlewared/middlewared/plugins/filesystem_/perm_check.py:30]: Meh, it's a private method, and the args are minimal.

**Comment by anodos325** [comment]
> Just some minor nits :)
> 
> Btw can we write some integration tests for this please ?

Yeah, I'll add some tomorrow.

**Review by sonicaj — APPROVED** [review]

File: tests/api2/test_190_filesystem.py [hunk 5]
```diff
@@ -277,3 +278,89 @@ def test_09_test_dosmodes():
...
+        2) group@ tag will test `gid` from payload
+        3) GROUP will test the `id` in payload with id_type
+        4) USER will test the `id` in mayload with USER id_type
+        """
+
+        # Start with testing denials
+        with create_dataset(sub_target, {'acltype': 'NFSV4', 'aclmode': 'PASSTHROUGH'}):
+            acl = deepcopy(NFSV4_DACL)
+            names = ['daemon', 'apps', 'nobody', 'nogroup']
+            for idx, entry in enumerate(NFSV4_DACL):
```
Comment (ericbsd) [inline tests/api2/test_190_filesystem.py:313]: From here I see indentation issues there is one extra space on almost all lines below.

File: tests/api2/test_190_filesystem.py [hunk 6]
```diff
@@ -277,3 +278,89 @@ def test_09_test_dosmodes():
...
+                     })
+                 elif entry['tag'] == 'GROUP':
+                     perm_job = POST('/filesystem/setacl/', {
+                         'path': path,
+                         'uid': 0,
+                         'gid': 0,
+                         'dacl': [entry]
+                     })
+
+                 job_status = wait_on_job(perm_job.json(), 180)
```
Comment (ericbsd) [inline tests/api2/test_190_filesystem.py:355]: wait_on_job is not imported.

**Review by ericbsd — CHANGES_REQUESTED** [review]

**Review by ericbsd — APPROVED** [review]
LGTM

**Pull request merged**